### DISCUSSION TITLE
delineate & summarize subgraphs by node attribute?

### DISCUSSION THREAD

**Discussion by fluviotect** [networkx/networkx#4586] [Q&A] [answered]
I've been tinkering with networkx for about a week and would like to use it for my current project.

My need is to iteratively subdivide the graph by changing node attributes, then summarize certain attributes (e.g. length, # of edges, et al.) by subdivision. I will have a large number of nodes and edges (likely tens of thousands) so need to automate. If it matters, my networks will have dendritic structure with an outlet I should always be able to specify. Sample graphic follows:

![testgraph_samp_4grps_scenarios_annotated](https://user-images.githubusercontent.com/41835370/106423435-50b05380-64c5-11eb-8949-7e4553eb3139.png)

I have a vision on how to proceed by hard-coding, but before I do, are does nx have native functions that will streamline my process? 

If I have to hard-code, can I accomplish the above with the structure in the attached graphic? Or, would I need to add a dummy node next to each 'off' node, insert 0-length edges (e.g. one in each direction as part of a MultiDiGraph) adjacent to each 'off' node and weight them, respectively, by whichever direction flux should or should not occur?

Cheers,
Will

**Comment by dschult** [comment]
I'm not sure what you mean by hard-code it because you haven't really described the decision process for this algorithm. You can certainly write algorithms to search for the next type 1 node group to split from the others and then enact that split by a method that depends on how you are storing and denoting your splittings. It is quite possible that literally splitting the graph into many small ones would be effective, but it is hard to say.  

If you do that and want to retain the larger scale connectivity you can make a large scale graph that literally uses smaller graphs as the nodes. It looks like you have two types of nodes. The type 1 nodes could be grouped into connected components by filtering out the type 2 nodes. Then the type 2 nodes could be represented in a large graph by an edge between groups. You might need to filter out type 2 nodes that are "off" or something. 

Of course, it is also possible that it is easier to just have one DiGraph with all the nodes and then use node attributes to group them, identify their type, etc. There is always a question about whether it is better to store nodeinfo in the DiGraph object or in a separate data structure (list of sets?).  Which is better depends on what you want to do with this info.

**Reply by fluviotect** [reply]
Cheers for the reply.

At my most granular level, I'm wondering if there is a way to simply pass a list of nodes IDs (e.g. Type 2 nodes that are 'off') to a networkx function that divides a pre-existing graph into subgraphs. For example...

`T2nodes = [n for n,v in J.nodes(data=True) if v['Type'] == 2]`

would return (for right panel in graphic):

`[12,13,14]`

This would streamline having to explicitly identify each subgraph element required by nx.subgraph().`

With regard to the more conceptual portion of my question, I've been trying the directed approach (DiGraph and MultiDiGraph) from your 3rd paragraph (where it seems like external storage is a bit better for my data and/or understanding of networkx). However, I'm liking the idea of your second paragraph and will give that a go once I figure an efficient way to generate the subgraphs.

**Reply by fluviotect** [reply]
I've figured-out something of a solution...if I use a DiGraph with only one set of edges directed away from my outlet, then I can iterate through T2nodes to generate subgraphs, then access edge attributes (e.g. length) by subgraph.  Given the anticipated size of my networks, I'm trying to be judicious with using too many FOR loops, but this runs OK on my test network.

**Comment by dschult** [comment, marked as answer]
The [GraphView features of Networkx](https://networkx.org/documentation/stable/reference/classes/index.html#module-networkx.classes.graphviews) allow you to easily hide a set of nodes or edges. It makes it easy for the user, but the hidden nodes and edges have to be rejected from consideration at every traversal. So it works great sometimes and is slow others. If it is slow, you can use `copy()` with `subgraph_view()` to easily construct the subgraphs as their own objects.

```python
cc_generator = nx.strongly_connected_components(nx.subgraph_view(G, node_filter=lambda n : G.nodes[n]["Type"] == 2))
lots_of_subgraphs = [G.subgraph(cc).copy() for cc in cc_generator]
```
(untested)

**Reply by fluviotect** [reply]
Much appreciated. Cheers!
